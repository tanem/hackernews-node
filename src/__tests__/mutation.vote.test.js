const path = require('path')
const bcrypt = require('bcryptjs')
const chance = require('chance').Chance(path.basename(__filename))
const jwt = require('jsonwebtoken')
const { APP_SECRET } = require('../utils')
const request = require('supertest')

let httpServerUrl
let prisma

beforeAll(async () => {
  httpServerUrl = global.__HTTP_SERVER_URL__
  prisma = global.__PRISMA__
})

const name = chance.first()
const email = chance.email()
const password = chance.word()
const description = chance.sentence({ words: 5 })
const url = chance.url()

let linkId
let token
let userId

beforeEach(async () => {
  ;({ id: userId } = await prisma.user.create({
    data: {
      name,
      email,
      password: await bcrypt.hash(password, 10),
    },
  }))
  ;({ id: linkId } = await prisma.link.create({
    data: {
      description,
      url,
      postedBy: {
        connect: { id: userId },
      },
    },
  }))

  token = jwt.sign({ userId }, APP_SECRET)
})

test('new vote', async () => {
  const { body, status } = await request(httpServerUrl)
    .post('/')
    .set('Authorization', `Bearer ${token}`)
    .send({
      query: `
        mutation {
          vote(linkId: "${linkId}") {
            id
            link {
              description
              id
              url
            }
            user {
              email
              id
              name
            }
          }
        }
      `,
    })

  expect(status).toBe(200)
  expect(body).toMatchSnapshot()
})

test('duplicate vote', async () => {
  await prisma.vote.create({
    data: {
      link: {
        connect: { id: linkId },
      },
      user: {
        connect: { id: userId },
      },
    },
  })

  const { body, status } = await request(httpServerUrl)
    .post('/')
    .set('Authorization', `Bearer ${token}`)
    .send({
      query: `
        mutation {
          vote(linkId: "${linkId}") {
            id
          }
        }
      `,
    })

  expect(status).toBe(200)
  expect(body).toMatchSnapshot()
})

test('not authenticated', async () => {
  const { body, status } = await request(httpServerUrl)
    .post('/')
    .send({
      query: `
        mutation {
          vote(linkId: "${linkId}") {
            id
            link {
              description
              id
              url
            }
            user {
              email
              id
              name
            }
          }
        }
      `,
    })

  expect(status).toBe(200)
  expect(body).toMatchSnapshot()
})
