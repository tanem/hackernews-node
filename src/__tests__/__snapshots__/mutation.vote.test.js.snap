// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`duplicate vote 1`] = `
Object {
  "data": Object {
    "vote": null,
  },
  "errors": Array [
    Object {
      "locations": Array [
        Object {
          "column": 11,
          "line": 3,
        },
      ],
      "message": "Already voted for link: 1",
      "path": Array [
        "vote",
      ],
    },
  ],
}
`;

exports[`new vote 1`] = `
Object {
  "data": Object {
    "vote": Object {
      "id": "1",
      "link": Object {
        "description": "Ceg nudde vokori joh zuffer.",
        "id": "1",
        "url": "http://kun.mu/pup",
      },
      "user": Object {
        "email": "jomu@huhubur.zm",
        "id": "1",
        "name": "Danny",
      },
    },
  },
}
`;

exports[`not authenticated 1`] = `
Object {
  "data": Object {
    "vote": null,
  },
  "errors": Array [
    Object {
      "locations": Array [
        Object {
          "column": 11,
          "line": 3,
        },
      ],
      "message": "Not authenticated",
      "path": Array [
        "vote",
      ],
    },
  ],
}
`;
